spicy/toolchain/src/compiler/codegen/parser-builder.cc

```diff
@@ -1021,7 +1022,32 @@ struct ProductionVisitor
                 auto pstate = pb->state();
                 pstate.literal_mode = mode;
                 pushState(std::move(pstate));
+
+                // Since `advance` can trigger recoverable errors when
+                // hitting a gap, bracket the call to it in a `try`/`catch`
+                // block if we are in search mode and attempt to recover.
+                if ( mode == LiteralMode::Search ) {
+                    auto [body, try_] = builder()->addTry();
+
+                    pushBuilder(try_.addCatch(builder::parameter(ID("e"), builder::typeByID("hilti::MissingData"))),
+                                [&]() {
+                                    // `advance` has failed, retry at the next non-gap block.
+                                    pb->advanceToNextData();
+
+                                    // // FIXME(bbannier):
+                                    // // We operate on `ncur` while `advanceToNextData`
+                                    // // updates `cur`; copy its result over.
+                                    // builder()->addAssign(ID("ncur"), state().cur);
+
+                                    // Continue incremental matching.
+                                    builder()->addContinue();
+                                });
+
+                    pushBuilder(body);
```

Comment:
Since we have this block of code twice now, I'd factor it out into a function, that will make this all more readable.

Reply:
The code for regexps needs a slight twist (accidentally left in here commented out and marked with `FIXME`), but we can capture that difference with an additional parameter. I added a helper in a new commit.